File: src/commands/client_side/RemoveHeadCommand.java
package commands.client_side;

import commands.client_side.core.ServerSideCommand;
import lib.ExecutionStatus;
import commands.client_side.core.ClientCallbackFunction;
import OrganizationDatabase.Organization;
import OrganizationDatabase.OrganizationManagerInterface;

public class RemoveHeadCommand extends ServerSideCommand {

    public RemoveHeadCommand(ClientCallbackFunction clientCallbackFunction, OrganizationManagerInterface organizationDatabase) {
        super(clientCallbackFunction, organizationDatabase);
    }

    @Override
    protected void executeImplementation(String[] args, ClientCallbackFunction callback) {
        Organization removedOrganization = organizationDatabase.removeHead();

        callback.invoke(
                removedOrganization != null ? ExecutionStatus.SUCCESS : ExecutionStatus.FAILURE,
                null,
                removedOrganization
        );
    }
}

File: src/commands/client_side/RemoveHeadCommand.kt
package commands.client_side

import commands.client_side.core.ServerSideCommand
import database.Organization
import database.OrganizationManagerInterface
import exceptions.OrganizationNotFoundException

class RemoveHeadCommand(
    organizationDatabase: OrganizationManagerInterface
) : ServerSideCommand(organizationDatabase) {
    override fun executeImplementation(argument: Any?): Result<Organization> {
        assert(argument == null)

        val removedOrganization = organizationDatabase.removeHead()

        if (removedOrganization == null) {
            return Result.failure(OrganizationNotFoundException())
        }

        return Result.success(removedOrganization)
    }
}
